### PR TITLE
Clean up Action class

### PR DESCRIPTION
**Pull request by PSUdaemon**

### CODE REVIEW

File: iocore/eventsystem/I_Action.h [hunk 1]
```diff
@@ -25,6 +25,10 @@
 #ifndef _I_Action_h_
 #define _I_Action_h_
 
+#define ACTION_RESULT_NONE ((Action *)(0x01))
+#define ACTION_RESULT_DONE ((Action *)(0x02))
+#define ACTION_IO_ERROR ((Action *)(0x04))
```
Comment (shinrich) [inline iocore/eventsystem/I_Action.h]: Is there any significance to changing the numbering for these constants?  I assume they aren't persistent, so it shouldn't make a difference.  Are you trying to make a bitmask?

**Comment by shinrich** [comment]
Overall looks like good changes.  Collapsing identical methods cancel and cancel action.  Eliminating unnecessary inline stuff.  Cleaning up a possible dangling action in HttpSM.  Once the compile errors are addressed, looks good to me.

File: iocore/eventsystem/I_Action.h (hunk 1 above)
Comment (PSUdaemon) [reply iocore/eventsystem/I_Action.h]: I'm really just trying to simplify the macros. Before it was setting it to values like 1,3,5 and 7. I'm not sure it has to be a bitmask, but that was simple to do incase it is used like that somewhere. It kinda looked like the old macros might have intended that, but weren't achieving it.

**Comment by PSUdaemon** [comment]
@shinrich turns out that inline stuff is needed for tests. It's pretty ugly. I feel like there must be a way around it, but out of scope for this PR.

**Comment by PSUdaemon** [comment]
[approve ci icc]

**Comment by PSUdaemon** [comment]
[approve ci autest]

**Comment by PSUdaemon** [comment]
[approve ci clang-analyzer]

**Comment by PSUdaemon** [comment]
[approve ci autest]

**Comment by PSUdaemon** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci debian]

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

File: iocore/eventsystem/I_Action.h (hunk 1 above)
Comment (SolidWallOfCode) [reply iocore/eventsystem/I_Action.h]: This should be `static constexpr` values inside the `Action` class. E.g.
```
static constexpr self_type * RESULT_NONE {reinterpret_cast<self_type*>(0x01)};
```

File: iocore/eventsystem/I_Action.h [hunk 2]
```diff
@@ -114,37 +116,11 @@ class Action
...
-    ink_assert(!c || c == continuation);
-#ifdef DEBUG
-    ink_assert(!cancelled);
-    cancelled = true;
-#else
-    if (!cancelled)
-      cancelled = true;
-#endif
-  }
+  bool cancelled = false;
```
Comment (SolidWallOfCode) [inline iocore/eventsystem/I_Action.h]: `std::atomic<bool>`?

File: iocore/eventsystem/I_Action.h [hunk 3]
```diff
@@ -158,16 +134,13 @@ class Action
...
-  cancel_action(Continuation *c = nullptr)
+  cancel(Continuation *c = nullptr)
   {
     ink_assert(!c || c == continuation);
-#ifdef DEBUG
     ink_assert(!cancelled);
-    cancelled = true;
-#else
-    if (!cancelled)
+    if (!cancelled) {
```
Comment (SolidWallOfCode) [inline iocore/eventsystem/I_Action.h]: Why not
```
ink_assert(!canceled);
canceled = true;
```

File: iocore/eventsystem/I_Action.h (hunk 1 above)
Comment (SolidWallOfCode) [reply iocore/eventsystem/I_Action.h]: I understand the previous macros, and that's not a bad idea. The goal is to make the pointers odd numbers, which are *always* invalid pointers. That makes them easy to distinguish from actual pointers. Note the macro shifted left and added on, i.e. opened bit 0 and set it, forcing the result to be odd.

**Comment by PSUdaemon** [comment]
I won't have time to make the necessary changes and validate them. I'm just going to close this and if someone like @SolidWallOfCode, @shinrich, or @zwoop wants to pick them up and make it better, they can.

✌️

**Pull request closed without merging**